inc/zoo/swar/SWAR.h

```diff
@@ -67,6 +67,21 @@ constexpr __uint128_t lsbIndex(__uint128_t v) noexcept {
 }
 #endif
 
+// This is placed ahead of other bare manipulation functions to reduce integer
+// promotion weirdness in the constant creation of SWAR and related.
+template<int NBits, typename T>
+constexpr auto leastNBitsMask() {
+    // It was integer promotion screwing this up at equal sizes all along, special casing unblocks.
+    constexpr auto type_bits = sizeof(T) * 8;
+    if constexpr(NBits < type_bits) {
+        return (T{1}<<NBits)-1;
+    } else if constexpr(type_bits < 32) {
+        return (T{1}<<NBits)-1;
+    } else if constexpr(type_bits >= 32) {
+        return ~T{0};
+    }
```

Comment:
It looks like I might be missing some context about this bit but is there an inconsistency with when/how different types will "saturate" if the NBits argument is bigger than the type has bits? 

```cpp
// Saturating behaviour for larger types.
static_assert(0xFFFFFFFFu == leastNBitsMask<64, u32>());
static_assert(0xFFFFFFFFFFFFFFFFu == leastNBitsMask<65, u64>());
static_assert(0xFFFFFFFFFFFFFFFFu == leastNBitsMask<420, u64>());

// We don't get the "saturating" effect on smaller types?
■ Static assertion failed due to requirement '65535U == leastNBitsMask<17, unsigned short>()'
static_assert(0xFFFFu == leastNBitsMask<17, u16>());

// Instead, we get this which passes the test:
static_assert(0x1FFFFu == leastNBitsMask<17, u16>());

```

Reply:
will resolve this lol